src/projector.cpp

```diff
@@ -152,22 +160,58 @@ void Projector::computeFourierTransformMap(
 		REPORT_ERROR("Projector::computeFourierTransformMap%%ERROR: Dimension of the data array should be 2 or 3");
 	}
 #ifdef CUDA
-	size_t mem_req;
+	static int semid = -1;
+	struct sembuf op_lock[2]=  { 0, 0, 0, /* wait for sem #0 to become 0 */
+	     						 0, 1, SEM_UNDO /* then increment sem #0 by 1 */ };
+	struct sembuf op_unlock[1]= { 0, -1, (IPC_NOWAIT | SEM_UNDO) /* decrement sem #0 by 1 (sets it to 0) */ };
+
+	size_t mem_req, ws_sz;
 	int Faux_sz = padoridim*(padoridim/2+1);
+	int n[3] = {padoridim, padoridim, padoridim};
+	cufftType cufft_type = CUFFT_R2C;
+	if(sizeof(RFLOAT) == sizeof(double))
+		cufft_type = CUFFT_D2Z;
 	
 	if(ref_dim == 3)
 		Faux_sz *= padoridim;
 
 	mem_req =  (size_t)1024;
 	if(do_heavy && do_gpu)
+	{
+	    cufftEstimateMany(ref_dim, n, NULL, 0, 0, NULL, 0, 0, cufft_type, 1, &ws_sz);
+
 		mem_req = (size_t)sizeof(RFLOAT)*MULTIDIM_SIZE(vol_in) +                   // dvol
 				  (size_t)sizeof(Complex)*Faux_sz +                                // dFaux
-				  (size_t)sizeof(RFLOAT)*MULTIDIM_SIZE(Mpad);                      // dMpad
+				  (size_t)sizeof(RFLOAT)*MULTIDIM_SIZE(Mpad) +                     // dMpad
+				  ws_sz + 16;                                                      // worspace for cuFFT
+	}
 
 	CudaCustomAllocator *allocator = NULL;
-	if(do_gpu) 
-		allocator = new CudaCustomAllocator(mem_req, (size_t)16);
+	if(do_gpu && do_heavy)
+	{
+		int devid;
+		size_t mem_free, mem_tot;
+		cudaDeviceProp devProp;
+		if (semid <0) 
+		{
+			HANDLE_ERROR(cudaGetDevice(&devid));
+			if ( ( semid=semget(SEMKEY+devid, 1, IPC_CREAT | PERMS )) < 0 ) 
+				REPORT_ERROR("semget error"); 
+		}
+		if (semop(semid, &op_lock[0], 2) < 0) 
+			REPORT_ERROR("semop lock error");
 
+		HANDLE_ERROR(cudaMemGetInfo(&mem_free, &mem_tot));
+		if(mem_free > mem_req)
+			allocator = new CudaCustomAllocator(mem_req, (size_t)16);
+		else
+		{
+			std::cerr << " WARNING: There is not enough memory to process the Expectation setup stage on GPU. It will be executed on CPU." << std::endl;
```

Comment:
Please remove this warning, because this is harmless (i.e. just use the old code path).